### DISCUSSION TITLE
3.08c Can't find slingbox

### DISCUSSION THREAD

**Discussion by Pthepeg** [GerryDazoo/Slinger#171] [Q&A] [answered]
First off, thanks Gerry for doing this project and sharing it. Great work. Looks like I can keep using my beloved 350 into the future. As soon as I got it going I went out and picked up another Slingbox350 for $20. Not sure what I'll do with it but happy to have it. I'm running 3.07 on a Raspberry Pi Model B Rev 2, and its been running very well.
Here is my issue, and I am stumped:
3.08c cannot find my Slingbox350.
3.07, using the very same config.ini, only a minute later, will find the 350 just fine.
Using 3.08c, if I put the slingbox IP:port into the config.ini, slingbox_server finds it and I can view the stream just fine.
Here are two output captures, first from the 3.08c and second from 3.07:

$ python3 -u slingbox_server.py
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 11958
Using config file config.ini
Connection Manager Running on port 62432 with 10 max streams using URL slingbox.
Streamer Running: starting up on  port 62432 8388608
 10 config.ini SLINGBOX 62432 62432 8388608
slingJim No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
slingJim Giving up. Sorry...

-----------------------------------------------------------------------------
$ python3 -u slingbox_server.py
Version : 3.07 Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 11977
Using config file config.ini
Connection Manager Running on port 62432 with 4 max streams....
Streamer Running: starting up on  port 62432
 4 config.ini 62432
No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
Slingbox Found 192.168.100.113 5202 " SlingJim "
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Using slingbox at  ('192.168.100.113', 5202)
Streamer:  slingJim Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

**Comment by Pthepeg** [comment]
I had a hunch that the slingbox wasn't being found because its on port 5202.
So I swapped out the slingbox for another 350 which I know is on port 5201.
I only power up one slingbox at a time.
Same problem, the slingbox_server will not find the slingbox unless I put the IP:port in the config.ini file.

**Comment by GerryDazoo** [comment]
How did you determine the IP address for the new 350?
Once you know the IP address and port numbers for each of your boxes just put that into your config.ini files and your done.

**Comment by Pthepeg** [comment]
Sure, in my router configuration I can see the slingboxes and allocate IP addresses statically to them and put those addresses in the config.ini files, and I'm done. Thanks.
Did I get the wrong impression from the 3.08 README.txt ?  I thought, like 3.07, if I include netifaces then I don't need to do the above. Cheers.

**Reply by GerryDazoo** [reply]
I'll review the document. You need netifaces if you need the server code to scan the local network for slingboxes. If you only have one box it will use the one it finds. If not you need to put the IP addresses into the config.ini(s) so it knows what to do.

**Reply by Pthepeg** [reply]
for my setup, 3.08 is not finding any boxes on its own. 3.07 does find the boxes.

**Reply by GerryDazoo** [reply]
Hmm, that doesn't sound right.  Could be a bug. Could you run a "tcpdump" to check to make sure the code is doing what I expect.
It sends out a broadcast message on the local LAN on port 5004. And then each of the singboxes should reply.

I don't have my Pi on had but I think this is the command you would run as root
tcpdump -i eth0 port 5004
I'm assuming the interface is the wired ethernet. replace with the interface your using if needed.

**Reply by Pthepeg** [reply]
(I only have one slingbox350 turned on and its not found)

$ sudo tcpdump -i eth0 port 5004
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
12:31:21.369542 IP slingserver.40817 > 192.168.100.255.5004: UDP, length 32
12:31:21.371312 IP Slingbox-350.5004 > slingserver.40817: UDP, length 124
^C
2 packets captured
2 packets received by filter
0 packets dropped by kernel
-------------------------------------------------------------------------------------------------
$ python3  slingbox_server.py configunified.ini
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 15913
Using config file configunified.ini
Connection Manager Running on port 62412 with 3 max streams using URL slingbox.
BOXES [('sb1', '350')]
Streamer Running:  3 configunified.ini 350 350 62412 8388608
starting up on  port 62412 8388608
350 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Building page for 350
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
350 Giving up. Sorry...
--------------------------------------------------------------------------------------------------
[SLINGBOXES]
sb1=350

[350]
sbtype="350/500/M1"
password=(i put the actual password here)
;finderid=BAD844CF70531865D73650A16A0A536
;ipaddress=192.168.100.113
;port=5202
Resolution=12
FrameRate=60
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=1
Remote=remote.txt

[SERVER]
; local port number for the server to listen on for connections
; port number+1 is used by server for internal communications.
port=62412
maxstreams=3
URLbase=slingbox

**Reply by GerryDazoo** [reply]
Well, this seems semi conclusive.
12:31:21.369542 IP slingserver.40817 > 192.168.100.255.5004: UDP, length 32
12:31:21.371312 IP Slingbox-350.5004 > slingserver.40817: UDP, length 124
The code is sending the broadcast and the slingbox is replying but your not getting a log.

Are you up for patching some python code? To help figure out what's happening.
Do you have a firewall running. Can you make sure UDP port 5004 is open.

**Reply by Pthepeg** [reply]
Sure, send me a patch. 
There is no firewall between the slingbox and slingserver, they are on the same LAN. Slingserver is on a pi, if there was any firewall causing this it should affect 3.07 too.

**Reply by GerryDazoo** [reply]
please uncomment (remove pound sign) from lines 132, 135 and 145.
That should tell us all we need to know. It's new code added in 3.08 to print out the returned FInderID as well as the slingbox name that's probably the culprit. I'm keen on seeing why it's works fine for me and my 3 boxes but not you.

**Reply by Pthepeg** [reply]
here you go Gerry:
------------------------------------------------------------
$ python3 slingbox_server_patch1.py configunified.ini
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 905
Using config file configunified.ini
Connection Manager Running on port 62412 with 3 max streams using URL slingbox.
BOXES [('sb1', '350')]
starting up on  port 62412 8388608
Streamer Running:  3 configunified.ini 350 350 62412 8388608
350 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
MSG 124 ('192.168.100.113', 5004)
(257, 0, 2, 0, 0, 0, 0, 0, 92, 0, 0, 0, 0, 0, 0, 0, 4, 104, 3, 4352, 58038, 22900, 214, 14523, 55276, 8825, 64833, 881, 83, 108, 105, 110, 103, 74, 105, 109, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5202, 6000)
Traceback (most recent call last):
  File "slingbox_server_patch1.py", line 140, in find_slingbox_info
    finderid = pbuf(finderid)[7:-2].replace(' ', '').upper()
  File "slingbox_server_patch1.py", line 85, in pbuf
    s = str(binascii.hexlify(s, ' ', 1))[2:-1].lower()
TypeError: hexlify() takes exactly one argument (3 given)
hexlify() takes exactly one argument (3 given) None
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Building page for 350
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server_patch1" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
350 Giving up. Sorry...

**Reply by GerryDazoo** [reply]
Interesting, Can you check your python3 version   python3 -version. I'm guessing it's an older release.
I think my code needs 3.5 or better.

For now to get you over this hump.
change line 139 to finderid = ''
and comment out line or delete line 140

**Reply by GerryDazoo** [reply]
Or try this, replace lines  134-143
with
                        port = msg[121] * 256 + msg[120]
                        name = ''
                        for char in msg[ 56:120]:
                              if char != 0: name = name + chr(char)
                        finderid = ''.join('{:02x}'.format(x) for x in msg[40:56])
                        print('Slingbox Found', source[0], port, '"',name, '"', 
                        'FinderID', finderid.upper())
                        boxes.append((source[0], port, name))

This should work with any 3.x Python version

**Reply by Pthepeg** [reply]
(see below)

**Comment by Pthepeg** [comment]
- python3 version is 3.73. I launch using **python3 slingbox_server.py** but the default was 2.7.16 so just to test it, I changed the default to python3 but get the same result with your patch1
- your patch2 on line 139,140 fixes the problem

I'm okay with this result (got my python default up to date :) but I don't mind at all if you want me to try some other patching/testing to get to the bottom of it.
 ----------------------- change python default--------------------------------------------
pi@slingserver:/home/slingbox $ python3 -V
Python 3.7.3
pi@slingserver:/home/slingbox $ **python -V
Python 2.7.16**
pi@slingserver:/home/slingbox $ update-alternatives --list python
update-alternatives: error: no alternatives for python
pi@slingserver:/home/slingbox $ sudo update-alternatives --install /usr/bin/python python /usr/bin/python2.7 1
update-alternatives: using /usr/bin/python2.7 to provide /usr/bin/python (python) in auto mode
pi@slingserver:/home/slingbox $ sudo update-alternatives --install /usr/bin/python python /usr/bin/python3.7 2
update-alternatives: using /usr/bin/python3.7 to provide /usr/bin/python (python) in auto mode
pi@slingserver:/home/slingbox $ **python -V
Python 3.7.3**
-----------------------your patch 1 result -----------------------------------------
$ python3 slingbox_server_patch1.py configunified.ini
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 1652
Using config file configunified.ini
Connection Manager Running on port 62412 with 3 max streams using URL slingbox.
BOXES [('sb1', 'ThreeFifty')]
starting up on  port 62412 8388608
Streamer Running:  3 configunified.ini ThreeFifty ThreeFifty 62412 8388608
ThreeFifty No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
MSG 124 ('192.168.100.113', 5004)
(257, 0, 2, 0, 0, 0, 0, 0, 92, 0, 0, 0, 0, 0, 0, 0, 4, 104, 3, 4352, 58038, 22900, 214, 14523, 55276, 8825, 64833, 881, 83, 108, 105, 110, 103, 74, 105, 109, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5202, 6000)
Traceback (most recent call last):
  File "slingbox_server_patch1.py", line 140, in find_slingbox_info
    finderid = pbuf(finderid)[7:-2].replace(' ', '').upper()
  File "slingbox_server_patch1.py", line 85, in pbuf
    s = str(binascii.hexlify(s, ' ', 1))[2:-1].lower()
TypeError: hexlify() takes exactly one argument (3 given)
hexlify() takes exactly one argument (3 given) None
**Can't find a slingbox on network.** Please make sure it's plugged in and connected. Check config.ini
Building page for ThreeFifty
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server_patch1" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
ThreeFifty Giving up. Sorry...

-----------------------your patch 2 result -----------------------------------------
$ python3 slingbox_server_patch2.py configunified.ini
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 1561
Using config file configunified.ini
Connection Manager Running on port 62412 with 3 max streams using URL slingbox.
BOXES [('sb1', 'ThreeFifty')]
starting up on  port 62412 8388608
Streamer Running:  3 configunified.ini ThreeFifty ThreeFifty 62412 8388608
ThreeFifty No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
MSG 124 ('192.168.100.113', 5004)
(257, 0, 2, 0, 0, 0, 0, 0, 92, 0, 0, 0, 0, 0, 0, 0, 4, 104, 3, 4352, 58038, 22900, 214, 14523, 55276, 8825, 64833, 881, 83, 108, 105, 110, 103, 74, 105, 109, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5202, 6000)
**Slingbox Found 192.168.100.113 5202** " SlingJim " FinderID
Building page for ThreeFifty
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server_patch2" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Traceback (most recent call last):
  File "slingbox_server_patch2.py", line 131, in find_slingbox_info
    msg, source = s.recvfrom(128)
socket.timeout: timed out
timed out None
ThreeFifty Using slingbox at  ('192.168.100.113', 5202)
Streamer:  ThreeFifty Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
10/27/2022, 12:57:15   Streaming connection from ('192.168.100.122', 56519)

**Reply by Pthepeg** [reply]
ok your patch3 (replace lines 134-143) works too.
------------------------------------------
$ python3 slingbox_server_patch3.py configunified.ini
Version : 3.08c Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 1676
Using config file configunified.ini
Connection Manager Running on port 62412 with 3 max streams using URL slingbox.
BOXES [('sb1', 'ThreeFifty')]
starting up on  port 62412 8388608
Streamer Running:  3 configunified.ini ThreeFifty ThreeFifty 62412 8388608
ThreeFifty No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Finding Slingbox on local network. My IP Info =  192.168.100.127
MSG 124 ('192.168.100.113', 5004)
**Slingbox Found 192.168.100.113 5202** " SlingJim " FinderID B6E27459D600BB38ECD7792241FD7103
Building page for ThreeFifty
Reading Custom Remote definition from remote.txt
 * Serving Flask app "slingbox_server_patch3" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Traceback (most recent call last):
  File "slingbox_server_patch3.py", line 131, in find_slingbox_info
    msg, source = s.recvfrom(128)
socket.timeout: timed out
timed out None
ThreeFifty Using slingbox at  ('192.168.100.113', 5202)
Streamer:  ThreeFifty Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

**Reply by GerryDazoo** [reply]
It is a python version issue. Hexilfy was changed in version 3.8: The sep and bytes_per_sep parameters were added.
Can you try that last patch I  gave to replace lines 134-143. If that works for you I'll incorporate it into the next release

**Reply by Pthepeg** [reply]
Yes I tried it, the results are above your comment. I called it patch3 and it does work on python V3.7.3

**Reply by GerryDazoo** [reply]
Many thanks for helping me sort this out.

**Reply by Pthepeg** [reply]
happy to help, good project.

**Comment by GerryDazoo** [comment, marked as answer]
Mark as answered

**Comment by GerryDazoo** [comment]
Fixed in 3.08d
